primitives/src/marketplace.rs

```diff
@@ -68,20 +76,72 @@ where
 		listing_fee: Option<CompoundFee<Balance>>,
 		account_list: Option<BoundedVec<AccountId, AccountSizeLimit>>,
 		offchain_data: Option<U8BoundedVec<OffchainDataLimit>>,
-	) -> MarketplaceData<AccountId, Balance, AccountSizeLimit, OffchainDataLimit> {
-		Self { owner, kind, commission_fee, listing_fee, account_list, offchain_data }
+		collection_list: Option<BoundedVec<CollectionId, CollectionSizeLimit>>,
+	) -> MarketplaceData<AccountId, Balance, AccountSizeLimit, OffchainDataLimit, CollectionSizeLimit>
+	{
+		Self {
+			owner,
+			kind,
+			commission_fee,
+			listing_fee,
+			account_list,
+			offchain_data,
+			collection_list,
+		}
 	}
 
-	pub fn allowed_to_list(&self, who: &AccountId) -> Option<()> {
+	// pub fn allowed_to_list(&self, who: &AccountId) -> Option<()> {
+	// 	let mut is_in_account_list = false;
+	// 	if let Some(account_list) = &self.account_list {
+	// 		is_in_account_list = account_list.contains(&who);
+	// 	}
+
+	// 	match self.kind {
+	// 		MarketplaceType::Public => !is_in_account_list,
+	// 		MarketplaceType::Private => is_in_account_list,
+	// 	}
+	// 	.then_some(())
+	// }
+
+	pub fn allowed_to_list(
+		&self,
+		who: &AccountId,
+		collection_id: Option<CollectionId>,
+	) -> Option<()> {
 		let mut is_in_account_list = false;
 		if let Some(account_list) = &self.account_list {
 			is_in_account_list = account_list.contains(&who);
 		}
 
-		match self.kind {
-			MarketplaceType::Public => !is_in_account_list,
-			MarketplaceType::Private => is_in_account_list,
+		if let Some(collection_id) = collection_id {
+			let mut is_in_collection_list = false;
```

Comment:
if you move this to line to line 115 then then the code will be simpler since you can merge the logic on lines 122, 123, 128 and 129.